src/components/page-header/image-panel-modal.tsx

```diff
@@ -0,0 +1,454 @@
+            reader.onerror = reject;
+            reader.readAsArrayBuffer(file);
+        });
+    }
+
+    async function sha256(bytes: Uint8Array): Promise<string> {
+        const hashBuffer = await crypto.subtle.digest('SHA-256', bytes);
+        return Array.from(new Uint8Array(hashBuffer))
+            .map(b => b.toString(16).padStart(2, '0'))
+            .join('');
+    }
+
+    const handleFileChange = async (event: React.ChangeEvent<HTMLInputElement>, tabId: number) => {
+        const type = tabId === 0 ? 'local' : 'server';
+        const file = event.target.files?.[0];
+        console.log('File selected:', file);
+        if (file) {
+            if (type === 'server') {
+                const bytes = await fileToBytes(file);
+                const hash = await sha256(bytes);
+                const reader = new FileReader();
+
+                const formData = new FormData();
+                formData.append('image', new Blob([bytes], { type: file.type }), file.name);
+                formData.append('mimetype', file.type);
+                formData.append('hash', hash);
+
+                fetch(image_endpoint, {
+                    method: 'POST',
+                    headers: {
+                        Authorization: `Bearer ${token}`,
+                    },
+                    body: formData,
+                })
+                    .then(response => response.json())
+                    .then(data => {
+                        if (!data.code || data.code === 201) {
+                            const imgId = `s${data.id}`;
+                            const newImage = {
+                                id: imgId,
+                                hash: data.hash,
+                                thumbnail: bytesToBase64DataURL(data.thumbnail, 'jpeg'),
+                            };
+                            setCloudList(prev => [...prev, newImage]);
+                            imageStoreIndexedDB.save(imgId, reader.result as string);
+                            setRefresh();
+                        } else {
+                            setError(data.message);
+                        }
+                    })
+                    .catch(error => console.error('Error uploading image:', error));
+            } else {
+                const reader = new FileReader();
+                reader.onload = () => {
+                    if (typeof reader.result === 'string') {
+                        const imgId = `l${nanoid(5)}`; // Use timestamp to ensure unique ID
+                        imageStoreIndexedDB.save(imgId, reader.result);
+                        setLocalList(prev => [...prev, { id: imgId, thumbnail: reader.result as string }]);
+                        setRefresh();
+                    }
+                };
+                reader.readAsDataURL(file);
+            }
+        }
+    };
+
+    async function fetchImageAsBase64(url: string): Promise<string> {
+        const response = await fetch(url, {
```

Comment:
pass id and construct url here, also send hash together with id to make sure no one could traverse all images